### ISSUE TITLE
Concurrency issue with the HTTP API

### ISSUE THREAD

**Issue by mikebryant** [OpenTSDB/opentsdb#306]
Sometimes when running concurrent requests, the response can go to a different request than the one it was intended for.

For example, running these two queries in parallel, I received the response for the second one from the first curl, while the second curl just hung:

/api/v1/tree/branch?branch=00015234374EB2ABF5C782F4928CB406977EB33BD3AE9FA1CD9E
/api/v1/tree/branch?branch=00015234374EB2ABF5C7

I got this exception and similar ones in the TSD logs:

```
14:49:23.952 INFO  [HttpQuery.logInfo] - [id: 0xdcc7d6e8, /192.168.254.102:9405 => /10.248.32.13:4242] HTTP /api/v1/tree/branch?branch=00015234374EB2ABF5C782F4928CB406977EB33BD3AE9FA1CD9E done in 238ms
14:49:23.953 ERROR [ConnectionManager.exceptionCaught] - Unexpected exception from downstream for [id: 0xdcc7d6e8, /192.168.254.102:9405 => /10.248.32.13:4242]
java.lang.AssertionError: This Deferred was already called!  New result=null, this=Deferred@1625036142(state=DONE, result=null, callback=<none>, errback=<none>)
        at com.stumbleupon.async.Deferred.callback(Deferred.java:982) ~[suasync-1.4.0.jar:fe17b98]
        at net.opentsdb.tsd.HttpQuery.done(HttpQuery.java:970) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.HttpQuery.sendBuffer(HttpQuery.java:1003) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.HttpQuery.sendReply(HttpQuery.java:802) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.TreeRpc.handleBranch(TreeRpc.java:252) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.TreeRpc.execute(TreeRpc.java:76) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.RpcHandler.handleHttpQuery(RpcHandler.java:255) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.RpcHandler.messageReceived(RpcHandler.java:163) ~[tsdb-1.9.0.jar:14fd1b1]
        at org.jboss.netty.handler.codec.http.HttpChunkAggregator.messageReceived(HttpChunkAggregator.java:148) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:107) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:312) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:88) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178) ~[netty-3.6.2.Final.jar:na]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110) [na:1.7.0_09]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603) [na:1.7.0_09]
        at java.lang.Thread.run(Thread.java:722) [na:1.7.0_09]
14:49:25.001 INFO  [HttpQuery.logInfo] - [id: 0x56c5ffdc, /192.168.254.102:14992 => /10.248.32.13:4242] HTTP /api/v1/tree/branch?branch=00015234374EB2ABF5C7 done in 157ms
14:49:25.153 INFO  [HttpQuery.logInfo] - [id: 0x56c5ffdc, /192.168.254.102:14992 => /10.248.32.13:4242] HTTP /api/v1/tree/branch?branch=00015234374EB2ABF5C7 done in 309ms
14:49:25.155 ERROR [ConnectionManager.exceptionCaught] - Unexpected exception from downstream for [id: 0x56c5ffdc, /192.168.254.102:14992 => /10.248.32.13:4242]
java.lang.AssertionError: This Deferred was already called!  New result=null, this=Deferred@1363779201(state=DONE, result=null, callback=<none>, errback=<none>)
        at com.stumbleupon.async.Deferred.callback(Deferred.java:982) ~[suasync-1.4.0.jar:fe17b98]
        at net.opentsdb.tsd.HttpQuery.done(HttpQuery.java:970) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.HttpQuery.sendBuffer(HttpQuery.java:1003) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.HttpQuery.sendReply(HttpQuery.java:802) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.TreeRpc.handleBranch(TreeRpc.java:252) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.TreeRpc.execute(TreeRpc.java:76) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.RpcHandler.handleHttpQuery(RpcHandler.java:255) ~[tsdb-1.9.0.jar:14fd1b1]
        at net.opentsdb.tsd.RpcHandler.messageReceived(RpcHandler.java:163) ~[tsdb-1.9.0.jar:14fd1b1]
        at org.jboss.netty.handler.codec.http.HttpChunkAggregator.messageReceived(HttpChunkAggregator.java:148) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:107) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:312) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:88) ~[netty-3.6.2.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178) ~[netty-3.6.2.Final.jar:na]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110) [na:1.7.0_09]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603) [na:1.7.0_09]
        at java.lang.Thread.run(Thread.java:722) [na:1.7.0_09]
14:49:29.167 INFO  [HttpQuery.logInfo] - [id: 0xa167fe51, /192.168.254.102:16181 => /10.248.32.13:4242] HTTP /api/v1/tree/branch?branch=00015234374EB2ABF5C7 done in 274ms
14:49:29.853 INFO  [HttpQuery.logInfo] - [id: 0x67bdc3d1, /192.168.254.102:16246 => /10.248.32.13:4242] HTTP /api/v1/tree/branch?branch=00015234374EB2ABF5C7 done in 272ms

```
